### PR TITLE
Reorg2

### PR DESCRIPTION
**Pull request by mrterry**
This builds on #464 and removes globals state from the provider registration process.

### CODE REVIEW

**Comment by mrterry** [comment]
@bollwyvl more refactoring to make writing tests easier.

**Comment by bollwyvl** [comment]
Looks good... but we may want to wait to merge/rebase this one until #464 is in... looks like a lot of duplicate changes.

**Comment by bollwyvl** [comment]
Ah, I see now that #464 is a strict subset, at least for the time being!

**Comment by mrterry** [comment]
Yeah. This branch presupposes the other.

**Comment by bollwyvl** [comment]
@mrterry thanks! please rebase this at your leisure... but I'll start it checking out!

File: nbviewer/handlers.py [hunk 1]
```diff
@@ -45,12 +47,25 @@ class CreateHandler(BaseHandler):
...
+    uri_rewrite_list = None
+
     def post(self):
         value = self.get_argument('gistnorurl', '')
-        redirect_url = transform_ipynb_uri(value, self.provider_rewrites)
+        redirect_url = transform_ipynb_uri(value, self.get_provider_rewrites())
         app_log.info("create %s => %s", value, redirect_url)
         self.redirect(redirect_url)
 
+    @classmethod
```
Comment (bollwyvl) [inline nbviewer/handlers.py]: good stuff. did something very similar over on ipymd, though I think it used a singleton instance.

File: nbviewer/utils.py [hunk 2]
```diff
@@ -53,40 +53,26 @@ def url_path_join(*pieces):
...
         if matches:
-            value = rewrite.format(*matches.groups())
-            break
+            return rewrite.format(*matches.groups())
 
     # encode query parameters as last url part
-    if '?' in value:
-        value, query = value.split('?', 1)
-        value = '%s/%s' % (value, quote('?' + query))
+    if '?' in uri:
```
Comment (bollwyvl) [inline nbviewer/utils.py]: Perhaps using urlparse here? considered doing this when i was working on the dropbox links. Not that this is wrong, and clearer.

File: nbviewer/handlers.py [hunk 3]
```diff
@@ -45,12 +47,25 @@ class CreateHandler(BaseHandler):
...
+    def get_provider_rewrites(cls):
+        # storing this on a class attribute is a little icky, but is better
+        # than the global this was refactored from.
+        if cls.uri_rewrite_list is None:
+            # providers is a list of module import paths
+            providers = self.settings['provider_rewrites']
```
Comment (bollwyvl) [inline nbviewer/handlers.py]: __cls__.settings!

Comment (bollwyvl) [inline nbviewer/handlers.py]: Ah, a bit more involved than that, since `settings` actually would only be on `self`. hmm...

Comment (bollwyvl) [inline nbviewer/handlers.py]: Perhaps this?
```python
    def get_provider_rewrites(self):
        # storing this on a class attribute is a little icky, but is better
        # than the global this was refactored from.
        if self.uri_rewrite_list is None:
            # providers is a list of module import paths
            providers = self.settings['provider_rewrites']
            setattr(self.__class__, "uri_rewrite_list",
                    provider_uri_rewrites(providers))
        return self.uri_rewrite_list
```

Comment (mrterry) [inline nbviewer/handlers.py]: i think

    type(self).uri_rewrite_list = providers

is easier to read.

**Comment by bollwyvl** [comment]
Looks really good! I'll press the green button tomorrow evening, if anyone else wants to take a look at it, but these are all positive changes in my eyes.

**Comment by bollwyvl** [comment]
Thanks again @mrterry! We really appreciate the contributions! Let me know how you want to proceed on #456. I'll also took a look at some of my outstanding stuff: it would be good to get ready to be able to cleanly move to the 4.0 line without a lot of PR debt!

**Comment by minrk** [comment]
Awesome, thanks both of you!

**Pull request merged**